File: tools/is2/fixture.py
#!/usr/bin/env python3
"""Source-owned IS2 diagnostics. No commercial import, installed software or GUI input.
Each run uses production --install, one dedicated cgroup and disposable prefix.
"""
import argparse,hashlib,json,os,pathlib,re,shutil,subprocess,time
CASES=('no_app','cooperative','ignores','exit_between','stale_pid','same_name','self','stale_mutex','stale_window','helper_zero','stale_recheck','access_denied','cancel_after','cleanup_preserves','powershell','wrapper1','wrapper3','exit23','contract')
def digest(p):
    with p.open('rb') as source:return hashlib.file_digest(source,'sha256').hexdigest()
def check_oracle(case, lines):
    rows=[]
    for line in lines:
        if line.startswith('IS2_CONTRACT'):continue
        match=re.fullmatch(r'IS2_API_V1 case=([a-z0-9_]+) stage=([a-z_]+) pid=([0-9]+) generation=([0-9]+) matches=(-?[0-9]+) outcome=([a-z0-9_]+) status=([0-9]+) tick=([0-9]+)',line)
        assert match and match[1]==case, 'malformed or cross-case oracle'
        rows.append(dict(zip(('case','stage','pid','generation','matches','outcome','status','tick'),match.groups())))
    def one(stage):
        found=[r for r in rows if r['stage']==stage];assert len(found)==1,(stage,found);return found[0]
    assert rows and rows[0]['stage']=='ready'
    if case in ('contract','exit23','wrapper1','wrapper3'):return {'schema':1,'case':case,'authority':'source_owned_fixture_only','result':'observed'}
    if case=='powershell':
        scripts=[r for r in rows if r['stage']=='script'];assert len(scripts)==4
        assert [r['outcome'] for r in scripts]==['probe0','probe1','probe2','probe3']
        sentinel=one('sentinel')
        return {'schema':1,'case':case,'authority':'source_owned_fixture_only','script_exit_statuses':[int(r['status']) for r in scripts], 'sentinel_created':sentinel['matches']=='1','requested_first_exit':37}
    present=one('presence');close=one('close')
    if case in ('no_app','stale_mutex'):
        assert present['matches']=='0' and close['outcome']=='not_requested' and one('recheck')['matches']=='0'
    elif case=='self':assert present['matches']=='1' and close['outcome']=='refused_identity'
    else:
        assert present['matches']=='1' and int(present['pid'])>0 and int(present['generation'])>0
        assert close['pid']==present['pid']
        recheck=one('recheck');assert (recheck['pid'],recheck['generation'])==(present['pid'],present['generation'])
        absent=case in ('cooperative','same_name','exit_between','stale_window','stale_recheck')
        assert recheck['matches']==('0' if absent else '1'),recheck
        expected=('refused_identity' if case in ('exit_between','stale_pid','stale_window') else 'access_denied' if case=='access_denied' else 'helper_zero_no_close' if case=='helper_zero' else 'request_accepted')
        assert close['outcome']==expected,close
        if case=='access_denied':assert int(close['status'])==5
        if case=='same_name':
            other=one('unrelated');assert other['matches']=='1' and other['pid']!=present['pid']
        if case=='stale_recheck':assert one('cached_recheck')['outcome']=='rejected_stale_snapshot'
        if case=='cancel_after':assert one('holding')['matches']=='1'
        else:assert one('cleanup')['matches']=='0'
    # Sanitized exact source-owned result. Never input to vendor classification.
    return {'schema':1,'case':case,'authority':'source_owned_fixture_only','presence_count':int(present['matches']),'close':close['outcome'],'close_status_domain':'win32_api' if case=='access_denied' else 'fixture_observation','close_status':int(close['status']),'recheck':[r['outcome'] for r in rows if r['stage']=='recheck'],'first_close_failure':case in ('ignores','stale_pid','self','stale_window','helper_zero','access_denied','cancel_after','cleanup_preserves'),'cause':'unestablished'}
def run(runtime,payload,adapter,output,case):
    os.umask(0o077);assert case in CASES;assert shutil.disk_usage(output.parent).free>2*1024**3
    output.mkdir(mode=0o700,parents=True,exist_ok=False)
    home=pathlib.Path.home();managed=home/'.local/share/linux-vst-bridge/managed';cli=home/'.local/bin/linux-vst-bridge'
    cap=json.loads(subprocess.check_output([str(cli),'capacity']))['capacity'];assert cap['dsp']==0 and cap['maintenance']==0 and not cap['cleanup_unconfirmed']
    runner=next(iter(json.loads((managed/'registry.json').read_text())['classes'].values()))['registration']['environment']['runner']
    root=output/'environment';root.mkdir(mode=0o700)
    for name in ('home','compatdata','runtime-var','host-cache','host-config','host-data','host-tmp','client'):(root/name).mkdir(mode=0o700)
    (root/'operation.lock').touch(mode=0o600);exe=output/'fixture.exe';shutil.copyfile(payload,exe);exe.chmod(0o400);exe.with_suffix('.exe.case').write_text(case+'\n')
    op=os.urandom(16).hex();report=output/'result.json';spec=output/'spec.json'
    value={'schema':2,'operation':op,'environment':{'id':op,'root':str(root),'revision':1,'runner':runner},'installer':{'path':str(exe),'sha256':digest(exe)},'format':'pe_executable','report':str(report)}
    if adapter:value['installer_launch']={'path':str(adapter),'sha256':digest(adapter)}
    spec.write_text(json.dumps(value));unit='linux-vst-bridge-installer-'+op+'.service';began=time.monotonic();cancel=False
    subprocess.run(['systemd-run','--user','--collect','--property=UMask=0077','--property=KillMode=control-group','--property=TimeoutStopSec=20','--property=StandardOutput=null','--property=StandardError=null','--unit='+unit,'/usr/bin/python3',str(runtime/'session.py'),'--install',str(spec)],check=True)
    try:
        while time.monotonic()-began<110:
            if report.exists():
                value=json.loads(report.read_text())
                if case=='cancel_after' and value.get('startup',{}).get('target_observation')=='observed' and time.monotonic()-began>18 and not cancel:
                    cancel=True;subprocess.run(['systemctl','--user','stop',unit],check=True,timeout=30)
                if value.get('cleanup_confirmed'):break
            time.sleep(.05)
    finally:
        if subprocess.run(['systemctl','--user','is-active','--quiet',unit]).returncode==0:
            stopped=subprocess.run(['systemctl','--user','stop',unit],timeout=30)
            if stopped.returncode:
                done=json.loads(report.read_text());assert done.get('cleanup_confirmed') and done.get('owned_live')==0 and subprocess.run(['systemctl','--user','is-active','--quiet',unit]).returncode!=0
    return finish(runtime,adapter,output,case,began,cancel)
def finish(runtime,adapter,output,case,began,cancel):
    root=output/'environment';exe=output/'fixture.exe';report=output/'result.json';op=json.loads((output/'spec.json').read_text())['operation']
    final=json.loads(report.read_text());private=json.loads((output/(op+'-transaction-private.json')).read_text());log=(output/(op+'-private.log')).read_bytes()
    events=[line.decode() for line in log.splitlines() if line.startswith((b'IS2_API_V1',b'IS2_CONTRACT'))]
    assert final['cleanup_confirmed'] and final['owned_live']==0
    survivors=[]
    for r in private['ledger']['processes']:
        try:
            if int((pathlib.Path('/proc')/str(r['pid'])/'stat').read_text().rsplit(')',1)[1].split()[19])==r['start_ticks']:survivors.append(r)
        except FileNotFoundError:pass
    assert not survivors
    contract=exe.with_suffix('.exe.contract')
    contract_hash=digest(contract) if contract.exists() else None
    runtime_images={}
    for arch in ('system32','syswow64'):
        image=root/'compatdata/pfx/drive_c/windows'/arch/'WindowsPowerShell/v1.0/powershell.exe'
        if image.exists():runtime_images[arch]={'sha256':digest(image),'size':image.stat().st_size}
    proof={'case':case,'contract_sha256':contract_hash,'powershell_images':runtime_images,'source':{n:digest(runtime/n) for n in ('session.py','ownership.py')},'adapter_sha256':digest(adapter) if adapter else None,'payload_sha256':digest(exe),'result':final,'fixture_api_oracle':events,'oracle_is_not_production_authority':True,'cancel_requested':cancel,'duration_seconds':time.monotonic()-began if began is not None else None,'same_identity_survivors':0}
    (output/'proof-private.json').write_text(json.dumps(proof,indent=2))
    # Cleanup is owned even when an attribution assertion fails; retain proof.
    assert root.resolve()==root;shutil.rmtree(root);proof['scratch_prefix_removed']=True;(output/'proof-private.json').write_text(json.dumps(proof,indent=2))
    if adapter:assert final['transaction']['launch_binding']['status']=='bound',final['transaction']['launch_binding']
    if case=='exit23':assert final['raw_exit']==23
    elif case=='cancel_after':assert final['state']=='cancelled'
    else:assert final['raw_exit']==0
    assert events,'source-owned payload did not report'
    proof['verified_oracle']=check_oracle(case,events)
    (output/'proof-private.json').write_text(json.dumps(proof,indent=2))
    print(json.dumps({'case':case,'outer':final['raw_exit'],'binding':final['transaction'].get('launch_binding'),'windows_roots':sum(r['target_root'] for r in private['windows_trace']['processes']),'presence_classes':final['transaction'].get('presence_close'),'cleanup':final['cleanup_confirmed'],'oracle':proof['verified_oracle']}))
if __name__=='__main__':
    p=argparse.ArgumentParser();p.add_argument('--runtime',type=pathlib.Path,required=True);p.add_argument('--payload',type=pathlib.Path,required=True);p.add_argument('--adapter',type=pathlib.Path);p.add_argument('--output',type=pathlib.Path,required=True);p.add_argument('--case',choices=CASES,required=True);a=p.parse_args();run(a.runtime.resolve(),a.payload.resolve(),a.adapter.resolve() if a.adapter else None,a.output.resolve(),a.case)

File: tools/is2/fixture.cpp
// Source-owned Windows API measurements. Output is a fixture oracle, not vendor
// role/ownership authority. The production supervisor observes roots and scripts.
#include <windows.h>
#include <cstdio>
#include <string>
#include <vector>
#include <array>
static std::wstring image(){wchar_t p[32768]{};DWORD n=GetModuleFileNameW(nullptr,p,32768);return {p,n};}
static unsigned long long born(HANDLE h){FILETIME c{},e{},k{},u{};if(!GetProcessTimes(h,&c,&e,&k,&u))return 0;return (static_cast<unsigned long long>(c.dwHighDateTime)<<32)|c.dwLowDateTime;}
static bool alive(HANDLE h){return WaitForSingleObject(h,0)==WAIT_TIMEOUT;}
static bool ignore_close=false;
static LRESULT CALLBACK window(HWND h,UINT m,WPARAM w,LPARAM l){if(m==WM_CLOSE){if(!ignore_close)DestroyWindow(h);return 0;}if(m==WM_DESTROY){PostQuitMessage(0);return 0;}return DefWindowProcW(h,m,w,l);}
static PROCESS_INFORMATION spawn(const std::wstring& args,const std::wstring& exe=L""){
 std::wstring cmd=L"\""+(exe.empty()?image():exe)+L"\" "+args;STARTUPINFOW s{};s.cb=sizeof(s);PROCESS_INFORMATION p{};
 if(!CreateProcessW(nullptr,cmd.data(),nullptr,nullptr,FALSE,0,nullptr,nullptr,&s,&p))return {};CloseHandle(p.hThread);return p;
}
static DWORD join(PROCESS_INFORMATION p,DWORD wait=15000){if(!p.hProcess)return 250;if(WaitForSingleObject(p.hProcess,wait)!=WAIT_OBJECT_0)return 251;DWORD code=252;GetExitCodeProcess(p.hProcess,&code);CloseHandle(p.hProcess);return code;}
struct Find {DWORD pid;HWND hwnd;unsigned count;};
static BOOL CALLBACK enumerate(HWND hwnd,LPARAM v){auto f=reinterpret_cast<Find*>(v);DWORD pid=0;GetWindowThreadProcessId(hwnd,&pid);wchar_t type[80]{};if(pid==f->pid&&GetClassNameW(hwnd,type,80)&&std::wstring(type)==L"IS2SourceOwned"){f->hwnd=hwnd;f->count++;}return TRUE;}
static HWND find(DWORD pid){Find f{pid,nullptr,0};EnumWindows(enumerate,reinterpret_cast<LPARAM>(&f));return f.count==1?f.hwnd:nullptr;}
static void emit(const std::string& test,const char* stage,DWORD pid,unsigned long long creation,int count,const char* outcome,DWORD status=0){
 std::printf("IS2_API_V1 case=%s stage=%s pid=%lu generation=%llu matches=%d outcome=%s status=%lu tick=%llu\n",test.c_str(),stage,pid,creation,count,outcome,status,GetTickCount64());std::fflush(stdout);
}
static bool close_exact(const std::string& test,PROCESS_INFORMATION p,HWND hwnd,unsigned long long expected,bool denied){
 DWORD owner=0;if(hwnd)GetWindowThreadProcessId(hwnd,&owner);
 if(p.dwProcessId==GetCurrentProcessId()||!p.hProcess||born(p.hProcess)!=expected||!alive(p.hProcess)||owner!=p.dwProcessId){emit(test,"close",p.dwProcessId,expected,-1,"refused_identity",0);return false;}
 if(denied){HANDLE limited=OpenProcess(PROCESS_QUERY_LIMITED_INFORMATION,FALSE,p.dwProcessId);BOOL result=limited?TerminateProcess(limited,29):FALSE;DWORD error=result?0:GetLastError();if(limited)CloseHandle(limited);emit(test,"close",p.dwProcessId,expected,1,result?"unexpected_success":"access_denied",error);return result!=FALSE;}
 SetLastError(0);BOOL sent=PostMessageW(hwnd,WM_CLOSE,0,0);DWORD error=sent?0:GetLastError();emit(test,"close",p.dwProcessId,expected,1,sent?"request_accepted":"request_refused",error);return sent!=FALSE;
}
static std::string case_file(){FILE* f=nullptr;_wfopen_s(&f,(image()+L".case").c_str(),L"rb");if(!f)return {};char b[80]{};auto n=fread(b,1,79,f);fclose(f);std::string s(b,n);while(!s.empty()&&(s.back()=='\n'||s.back()=='\r'))s.pop_back();return s;}
int wmain(int argc,wchar_t** argv){
 if(argc==2&&std::wstring(argv[1])==L"--self-test")return born(GetCurrentProcess())?0:1;
 if(argc==3&&std::wstring(argv[1])==L"--app"){
  ignore_close=std::wstring(argv[2])==L"ignore";WNDCLASSW c{};c.lpfnWndProc=window;c.hInstance=GetModuleHandleW(nullptr);c.lpszClassName=L"IS2SourceOwned";if(!RegisterClassW(&c))return 241;
  HWND h=CreateWindowW(c.lpszClassName,L"IS2 source-owned",WS_OVERLAPPED,0,0,100,100,nullptr,nullptr,c.hInstance,nullptr);if(!h)return 242;
  MSG m{};while(GetMessageW(&m,nullptr,0,0)>0){TranslateMessage(&m);DispatchMessageW(&m);}return 0;
 }
 if(argc==3&&std::wstring(argv[1])==L"--wrap"){auto n=wcstoul(argv[2],nullptr,10);return static_cast<int>(join(spawn(n?L"--wrap "+std::to_wstring(n-1):L"--payload")));}
 if(argc==5&&std::wstring(argv[1])==L"--adapter"){
  auto n=wcstoul(argv[2],nullptr,10);if(n>3)return 248;
  std::wstring args=n?L"--adapter "+std::to_wstring(n-1)+L" \""+argv[3]+L"\" \""+argv[4]+L"\"":L"\""+std::wstring(argv[4])+L"\"";
  return static_cast<int>(join(spawn(args,n?L"":std::wstring(argv[3]))));
 }
 std::string test=case_file();if(test.empty())return 240;
 emit(test,"ready",GetCurrentProcessId(),born(GetCurrentProcess()),-1,"source_owned");
 if(argc==2&&std::wstring(argv[1])==L"--payload")return 0;
 if(test=="wrapper1"&&argc==1)return static_cast<int>(join(spawn(L"--wrap 0")));
 if(test=="wrapper3"&&argc==1)return static_cast<int>(join(spawn(L"--wrap 2")));
 if(test=="exit23")return 23;
 if(test=="contract"){
  wchar_t cwd[32768]{};GetCurrentDirectoryW(32768,cwd);
  FILE* file=nullptr;_wfopen_s(&file,(image()+L".contract").c_str(),L"wb");if(!file)return 246;
  std::wstring data=L"cwd="+std::wstring(cwd)+L"\n";
  for(auto key:{L"HOME",L"USER",L"TEMP",L"PATH",L"WINEDLLOVERRIDES"}){wchar_t v[32768]{};GetEnvironmentVariableW(key,v,32768);data+=std::wstring(key)+L"="+v+L"\n";}
  HANDLE token=nullptr;TOKEN_ELEVATION elevation{};DWORD length=0;
  if(OpenProcessToken(GetCurrentProcess(),TOKEN_QUERY,&token)){GetTokenInformation(token,TokenElevation,&elevation,sizeof(elevation),&length);CloseHandle(token);}
  data+=L"elevated="+std::to_wstring(elevation.TokenIsElevated)+L"\n";fwrite(data.data(),sizeof(wchar_t),data.size(),file);fclose(file);
  std::printf("IS2_CONTRACT retained=private\n");std::fflush(stdout);return 0;
 }
 if(test=="powershell"){
  std::vector<std::wstring> scripts={L"[IO.File]::WriteAllText('C:\\IS2-script-sentinel.txt', 'source_owned'); exit 37",L"if (Get-Command Get-CimInstance -ErrorAction SilentlyContinue) { exit 0 } else { exit 1 }",L"if ((Get-CimInstance -ClassName Win32_Process | ? {$_.Path -and $_.Path.StartsWith('C:\\IS2-No-Application-Here', 'CurrentCultureIgnoreCase')}).Count -gt 0) { exit 0 } else { exit 1 }",L"Get-CimInstance -ClassName Win32_Process | ? {$_.Path -and $_.Path.StartsWith('C:\\IS2-No-Application-Here', 'CurrentCultureIgnoreCase')} | % { Stop-Process -Id $_.ProcessId -Force }"};
  wchar_t sys[32768]{};GetSystemDirectoryW(sys,32768);std::wstring ps=std::wstring(sys)+L"\\WindowsPowerShell\\v1.0\\powershell.exe";
  unsigned idx=0;for(auto& script:scripts){auto child=spawn(L"-C \""+script+L"\"",ps);DWORD pid=child.dwProcessId;auto generation=child.hProcess?born(child.hProcess):0;DWORD code=join(child);emit(test,"script",pid,generation,-1,("probe"+std::to_string(idx++)).c_str(),code);}
  emit(test,"sentinel",0,0,GetFileAttributesW(L"C:\\IS2-script-sentinel.txt")!=INVALID_FILE_ATTRIBUTES?1:0,"script_file_side_effect");return 0;
 }
 if(test=="no_app"||test=="stale_mutex"){
  HANDLE mutex=test=="stale_mutex"?CreateMutexW(nullptr,FALSE,L"Local\\IS2SourceOwnedStale"):nullptr;
  emit(test,"presence",0,0,0,mutex?"mutex_without_process":"absent");emit(test,"close",0,0,0,"not_requested");emit(test,"recheck",0,0,0,"absent");if(mutex)CloseHandle(mutex);return 0;
 }
 if(test=="self"){
  emit(test,"presence",GetCurrentProcessId(),born(GetCurrentProcess()),1,"self_match");PROCESS_INFORMATION self{};self.dwProcessId=GetCurrentProcessId();self.hProcess=GetCurrentProcess();close_exact(test,self,nullptr,born(self.hProcess),false);return 0;
 }
 bool ignore=test=="ignores"||test=="helper_zero"||test=="cancel_after"||test=="cleanup_preserves";
 auto child=spawn(ignore?L"--app ignore":L"--app cooperative");if(!child.hProcess)return 243;
 auto generation=born(child.hProcess);HWND hwnd=nullptr;
 for(unsigned k=0;k<200&&!hwnd;k++){hwnd=find(child.dwProcessId);Sleep(10);}
 if(!hwnd){TerminateProcess(child.hProcess,244);join(child);return 244;}
 PROCESS_INFORMATION unrelated{};if(test=="same_name"){unrelated=spawn(L"--app ignore");if(!unrelated.hProcess)return 245;}
 emit(test,"presence",child.dwProcessId,generation,1,"exact_handle_and_window_owner");
 if(test=="exit_between"||test=="stale_window"){PostMessageW(hwnd,WM_CLOSE,0,0);WaitForSingleObject(child.hProcess,5000);}
 if(test=="helper_zero")emit(test,"close",child.dwProcessId,generation,1,"helper_zero_no_close",0);
 else close_exact(test,child,hwnd,test=="stale_pid"?generation+1:generation,test=="access_denied");
 WaitForSingleObject(child.hProcess,500);bool remains=alive(child.hProcess);
 emit(test,"recheck",child.dwProcessId,generation,remains?1:0,remains?"still_present":"exited");
 if(test=="stale_recheck")emit(test,"cached_recheck",child.dwProcessId,generation,1,"rejected_stale_snapshot");
 if(unrelated.hProcess){emit(test,"unrelated",unrelated.dwProcessId,born(unrelated.hProcess),alive(unrelated.hProcess)?1:0,"preserved");TerminateProcess(unrelated.hProcess,0);join(unrelated);}
 if(test=="cancel_after"){emit(test,"holding",child.dwProcessId,generation,remains?1:0,"await_owned_cancellation");Sleep(180000);}
 if(remains)TerminateProcess(child.hProcess,0);join(child);
 emit(test,"cleanup",0,0,0,"source_owned_children_retired");return 0;
}
